src/game/assets.rs

```diff
@@ -44,58 +50,64 @@ pub enum SfxAsset {
     Step4,
 }
 
-#[derive(Resource, Reflect, Deref, DerefMut)]
-pub struct SfxAssets(HashMap<SfxAsset, Handle<AudioSource>>);
-
-impl SfxAssets {
-    pub fn new(asset_server: &AssetServer) -> Self {
-        let mut assets = HashMap::new();
-
-        assets.insert(
-            SfxAsset::ButtonHover,
-            asset_server.load("audio/sfx/button_hover.ogg"),
-        );
-        assets.insert(
-            SfxAsset::ButtonPress,
-            asset_server.load("audio/sfx/button_press.ogg"),
-        );
-        assets.insert(SfxAsset::Step1, asset_server.load("audio/sfx/step1.ogg"));
-        assets.insert(SfxAsset::Step2, asset_server.load("audio/sfx/step2.ogg"));
-        assets.insert(SfxAsset::Step3, asset_server.load("audio/sfx/step3.ogg"));
-        assets.insert(SfxAsset::Step4, asset_server.load("audio/sfx/step4.ogg"));
+impl AssetKey for SfxKey {
+    type Asset = AudioSource;
 
-        Self(assets)
-    }
-
-    pub fn all_loaded(&self, assets: &Assets<AudioSource>) -> bool {
-        self.0.iter().all(|(_, handle)| assets.contains(handle))
+    fn load(asset_server: &AssetServer) -> AssetMap<Self> {
+        AssetMap(HashMap::from([
+            (
+                SfxKey::ButtonHover,
+                asset_server.load("audio/sfx/button_hover.ogg"),
+            ),
+            (
+                SfxKey::ButtonPress,
+                asset_server.load("audio/sfx/button_press.ogg"),
+            ),
+            (SfxKey::Step1, asset_server.load("audio/sfx/step1.ogg")),
+            (SfxKey::Step2, asset_server.load("audio/sfx/step2.ogg")),
+            (SfxKey::Step3, asset_server.load("audio/sfx/step3.ogg")),
+            (SfxKey::Step4, asset_server.load("audio/sfx/step4.ogg")),
+        ]))
     }
 }
 
 #[derive(PartialEq, Eq, Hash, Reflect)]
-pub enum SoundtrackAsset {
+pub enum SoundtrackKey {
     Credits,
     Gameplay,
 }
 
-#[derive(Resource, Reflect, Deref, DerefMut)]
-pub struct SoundtrackAssets(HashMap<SoundtrackAsset, Handle<AudioSource>>);
+impl AssetKey for SoundtrackKey {
+    type Asset = AudioSource;
 
-impl SoundtrackAssets {
-    pub fn new(asset_server: &AssetServer) -> Self {
-        let mut assets = HashMap::new();
-        assets.insert(
-            SoundtrackAsset::Credits,
-            asset_server.load("audio/soundtracks/Monkeys Spinning Monkeys.ogg"),
-        );
-        assets.insert(
-            SoundtrackAsset::Gameplay,
-            asset_server.load("audio/soundtracks/Fluffing A Duck.ogg"),
-        );
-        Self(assets)
+    fn load(asset_server: &AssetServer) -> AssetMap<Self> {
+        AssetMap(HashMap::from([
+            (
+                SoundtrackKey::Credits,
+                asset_server.load("audio/soundtracks/Monkeys Spinning Monkeys.ogg"),
+            ),
+            (
+                SoundtrackKey::Gameplay,
+                asset_server.load("audio/soundtracks/Fluffing A Duck.ogg"),
+            ),
+        ]))
     }
+}
+
+pub trait AssetKey: Sized {
```

Comment:
I get why this is called `Key`, but it highlights the implementation detail and not the intention behind it. Also, the function `load` doesn't map semantic sense on a key and implies this is conceptually the asset itself. Maybe something like `PreloadedAsset` would be nicer?

Reply:
@paul-hansen do you have an opinion here?

Reply:
I think that, fundamentally, actual keys are not a thing that loads anything

Reply:
This is going to be a tradeoff between conceptual purity and code quality.

Reply:
I pushed a potential workaround for this, removing `AssetKey::load` in favor of manual impls of `FromWorld` for `AssetMap<XyzKey>`. It adds some boilerplate for each new asset collection, but you get the conceptual purity.

Reply:
My argument stems from the idea that `Key` is a bit of an implementation detail anyways and does not help you conceptualize what is happening, hence `PreloadedAsset`.
I'm fine with leaving it like this if you disagree, especially considering @paul-hansen's endorsement as an unbiased third party.

Reply:
I'd argue that the preloaded part is the implementation detail that the rest of the codebase doesn't need to know or care about. :P